pkg/collector/corechecks/snmp/discovery/discovery.go

```diff
@@ -0,0 +1,294 @@
+type Device struct {
+	entityID    string
+	deviceIP    string
+	deviceCheck *devicecheck.DeviceCheck
+}
+type snmpSubnet struct {
+	config     *checkconfig.CheckConfig
+	startingIP net.IP
+	network    net.IPNet
+
+	cacheKey string
+	devices  map[string]string
+
+	deviceFailures map[string]int
+}
+
+type snmpJob struct {
+	subnet    *snmpSubnet
+	currentIP net.IP
+}
+
+// Start discovery
+func (d *Discovery) Start() {
+	log.Debugf("subnet %s: Start discovery", d.config.Network)
+	go d.checkDevices()
+}
+
+// Stop signal discovery to shut down
+func (d *Discovery) Stop() {
+	log.Debugf("subnet %s: Stop discovery", d.config.Network)
+	close(d.stop)
+}
+
+// GetDiscoveredDeviceConfigs returns discovered device configs
+func (d *Discovery) GetDiscoveredDeviceConfigs() []*devicecheck.DeviceCheck {
+	d.discDevMu.RLock()
+	defer d.discDevMu.RUnlock()
+
+	discoveredDevices := make([]*devicecheck.DeviceCheck, 0, len(d.discoveredDevices))
+	for _, device := range d.discoveredDevices {
+		discoveredDevices = append(discoveredDevices, device.deviceCheck)
+	}
+	return discoveredDevices
+}
+
+// Start discovery
+func (d *Discovery) runWorker(w int, jobs <-chan snmpJob) {
+	log.Debugf("subnet %s: Start SNMP worker %d", d.config.Network, w)
+	for {
+		select {
+		case <-d.stop:
+			log.Debugf("subnet %s: Stop SNMP worker %d", d.config.Network, w)
+			return
+		case job := <-jobs:
+			log.Debugf("subnet %s: Handling IP %s", d.config.Network, job.currentIP.String())
+			err := d.checkDevice(job)
+			if err != nil {
+				log.Errorf(err.Error())
+			}
+		}
+	}
+}
+
+func (d *Discovery) checkDevices() {
+	ipAddr, ipNet, err := net.ParseCIDR(d.config.Network)
+	if err != nil {
+		log.Errorf("subnet %s: Couldn't parse SNMP network: %s", d.config.Network, err)
+		return
+	}
+
+	startingIP := ipAddr.Mask(ipNet.Mask)
+
+	configHash := d.config.DiscoveryDigest(d.config.Network)
+	cacheKey := fmt.Sprintf("%s:%s", cacheKeyPrefix, configHash)
+
+	subnet := snmpSubnet{
+		config:         d.config,
+		startingIP:     startingIP,
+		network:        *ipNet,
+		cacheKey:       cacheKey,
+		devices:        map[string]string{},
+		deviceFailures: map[string]int{},
+	}
+
+	d.loadCache(&subnet)
+
+	jobs := make(chan snmpJob)
+	for w := 0; w < d.config.DiscoveryWorkers; w++ {
+		go d.runWorker(w, jobs)
+	}
+
+	discoveryTicker := time.NewTicker(time.Duration(d.config.DiscoveryInterval) * time.Second)
+
+	for {
+		log.Debugf("subnet %s: Run discovery", d.config.Network)
+		startingIP := make(net.IP, len(subnet.startingIP))
+		copy(startingIP, subnet.startingIP)
+		for currentIP := startingIP; subnet.network.Contains(currentIP); incrementIP(currentIP) {
+
+			if ignored := subnet.config.IsIPIgnored(currentIP); ignored {
+				continue
+			}
+
+			jobIP := make(net.IP, len(currentIP))
+			copy(jobIP, currentIP)
+			job := snmpJob{
+				subnet:    &subnet,
+				currentIP: jobIP,
+			}
+			jobs <- job
+
+			select {
```

Comment:
Opinionated, I would move this at the beginning of the loop, line `128` to early return if stopped

Reply:
I think it doesn't change how earlier it will return since discovery can be stopped at any moment and we don't know where we will be in the loop when it stops.